### ISSUE TITLE
Add Swift Package Manager support

### ISSUE THREAD

**Issue by ChrisPearce** [AzureAD/microsoft-authentication-library-for-objc#737]
It would be very helpful if Swift Package Manager was supported by including a Package.swift file.

**Comment by pascalfriedrich** [comment]
Is there a rough timetable when the support will be added?

**Comment by jboisjo** [comment]
We need this badly as well :(

**Comment by glatour** [comment]
I can't wait to have the library moved to the swift package manager!!!

**Comment by mitchdenny** [comment]
Any update on plans for supporting Swift Package Manager. We are taking a dependency on this library and would like to support folks using Swift PM.

**Comment by kincade-dev** [comment]
Please add this!

**Comment by oldalton** [comment]
We could use some community help with this one :) I did some experimentation and it seems difficult to get it working considering we're heavily using prefix files and nested folders. Would appreciate any references of a successful example of SPM for Objective-C library without refactoring project structure, direct contributions or examples of getting header search working with nested folders.

**Comment by VladIacobIonut** [comment]
@oldalton Hi, I think Realm could be a good example of a library which is available via SPM and is written in Obj-C also. I hope it helps.

**Comment by kbartyzel** [comment]
Any update on this? This is critical

**Comment by nidegen** [comment]
@oldalton nested folders should not be a problem, but what do you mean with prefix files?

**Comment by nidegen** [comment]
@brandwe @rohitnarula7176 I would put that on top of your list, makes life much easier for many;)

**Comment by oldalton** [comment]
@nidegen, thanks for providing your input. Can you give more details about getting Swift package manager to work with nested folders? The only way I got it to work is to declare all paths separately which is very difficult to maintain as well as symlinking public headers (see some experimentation here: https://github.com/AzureAD/microsoft-authentication-library-common-for-objc/compare/dev...oldalton/swift_package_manager_support?expand=1). For CocoaPods we just use wildcards to resolve nested folders, but there doesn't seem to be the same way for Swift package manager. 

If you have some ideas how to make SPM support a little bit more maintainable, or you would like to contribute a PR from your side, your help/advice would be appreciated :)

**Comment by nidegen** [comment]
Yes I started a fork but declaring all nested paths. I think that is the only way currently. 
- Is that a no-go? Do the paths change that often?
-Reorganize the nested folders is not an option?

**Comment by BenLeggiero** [comment]
Thanks, @nidegen! In my experience, that's acceptable - just add another item to the release checklist to make sure these are in-sync

**Comment by oldalton** [comment]
@nidegen, I think even after adding all paths I still couldn't make it to work (as you can see in my branch). If you can get it to work for MSAL, we'd welcome a PR in our main repository such as other people could also benefit from it :)

**Comment by nidegen** [comment]
@oldalton yea I saw your branch. Quite some changes. Lets see!

**Comment by BenLeggiero** [comment]
Any progress on this? I'm about to integrate MSAL into our codebase and would love to do that as a Swift package!

**Comment by oldalton** [comment]
@BenLeggiero, we're still looking forward for an external help/contribution on this one! We've tried a few more options, but seems that SPM doesn't work very well with submodules either.

**Comment by BenLeggiero** [comment]
Thanks for the update, @oldalton. I might try my hand at it sometime later. Sad to hear SPM and Git Submodules aren't friends; hopefully that changes soon.

**Comment by bsiegel** [comment]
A quick update - Apple [announced](https://developer.apple.com/videos/play/wwdc2020/10147/) during WWDC this year that SwiftPM will now [support binary framework packages](https://developer.apple.com/documentation/swift_packages/distributing_binary_frameworks_as_swift_packages), providing a way forward for projects that use native code, ObjC, or use structures that are not conducive to shipping as a native source-based SwiftPM package, to onboard to SwiftPM.

This means that in order to support SwiftPM, the MSAL team should only need to compile the library into an .xcframework, and provide a Package.swift metadata file in the repo for SwiftPM. Once that's done people will be able to consume the dependency via SwiftPM. The project is already structured to compile into a framework so it shouldn't require changes to the project structure. However the MSAL team doesn't distribute pre-compiled copies today (the releases are only project sources), so they would need to begin building it into frameworks (and then combining those into a meta xcframework) as part of the release process.

**Comment by mitchdenny** [comment]
Following on what @bsiegel mentioned. I did some experiments building the various .framework files using xcodebuild and then creating the xcframework file. It all works fine with the code base as it stands at the moment. All that would be required is the addition of a Package.swift file to the repository that points to the location of that binary framework.

My suggestion would be that you add the steps to produce the .framework files and the .xcframework files to your build process and so when you tag your repository you upload the associated .xcframework. The Package.swift file would be updated prior to creating the tag so that it contains a pointer to the location of the .xcframework file in GitHub releases. At which point I think it should all just work.

**Comment by mitchdenny** [comment]
OK ... so using _Xcode 12 beta_ I was able to setup a build of the MSAL library as a binary framework, downloadable via Swift Package Manager. I built the .xcframework locally (just iOS and iOS simulator frameworks) and uploaded it to a GitHub release in my fork of the repo. I then created a ```Package.swift``` file which points to the release download location (and computed the checksum).

Once that was done I was able to download the package directly from that branch:

![Screen Shot 2020-07-04 at 4 45 57 pm](https://user-images.githubusercontent.com/513398/86508103-b863ca80-be20-11ea-92bd-f9187689ed7d.png)

![Screen Shot 2020-07-04 at 5 03 50 pm](https://user-images.githubusercontent.com/513398/86508107-bbf75180-be20-11ea-8cba-0af1c258d77a.png)

You can see that it cloned the repo and then pulled down the xcframework file:

![Screen Shot 2020-07-04 at 5 04 17 pm](https://user-images.githubusercontent.com/513398/86508109-be59ab80-be20-11ea-8db2-e81158272075.png)

The symbols are also showing up in Xcode.

![Screen Shot 2020-07-04 at 5 10 02 pm](https://user-images.githubusercontent.com/513398/86508110-c0bc0580-be20-11ea-9325-da746549ed57.png)

I haven't done any further experimentation yet to prove that the library works as advertised. But its looking pretty promising.

**Comment by sharmaniti** [comment]
@mitchdenny  - I created .xcframework from MSAL.framework and import it locally in my project to test. But i'm getting build error  as
'
Showing All Messages
While building for iOS Simulator, no library for this platform was found in 'MSAL.xcframework'.
'
simulator selected iPhone11 13.7

Any help to resolve this ?

**Comment by ryanthon** [comment]
Any updates on making this work?

**Comment by mitchdenny** [comment]
An ```*.xcframework``` is a directory layout which contains multiple ```*.framework``` outputs. It sounds tome like you may have built a ```.framework``` for a platform other than your intended target. The examples I've seen from Apple show packaging multiple ```.framework``` dirs into a single ```.xcframework```.

**Comment by sharmaniti** [comment]
@mitchdenny  - IS it possible for you to share `package.swift` file for reference please ?

**Comment by mitchdenny** [comment]
Here is an example here:

https://github.com/mitchdenny/microsoft-authentication-library-for-objc/blob/swift-pm-via-binary-framework/Package.swift

This was a while ago so the package description format could have been tweaked in the meantime.

**Comment by sharmaniti** [comment]
@mitchdenny - Thanks, solution works for me only on simulator not on devices. App crashes on launch  `dyld: Library not loaded: @rpath/MSAL.framework/` and when i use same .xcframework file directly in xcode project, just drag and drop in link libraries it works on devices as well as simulator but not as a swift package.

Not sure is there any way to disable code signing for .xcframework, when added as a swift package.
Here is the thread i found https://developer.apple.com/forums/thread/652873

**Comment by mitchdenny** [comment]
Yeah -- I think I signed that ```.xcframework``` with my local Apple developer ID rather than anything that you could verify. If you clone the repo and setup your own Package.swift file and build and host the .xcframework yourself, you should be able to get it going end to end, but obviously the AAD team is going to need to produce an official .xcframework if they want to support Swift Package Manager.

**Comment by schlingding** [comment]
Any updates? This is halting our adoption of an Azure AD in our new applications.

**Comment by Limoo** [comment]
We would also like to convert to Swift Package Manager. MSAL is the only framework that is still missing support

**Comment by draskovits** [comment]
Also interested in SPM support. I tried to include it with a xcframework, and it seems to work. There is currently an issue with SPM/Xcode when trying to use a binary dependency in a swift package (https://bugs.swift.org/browse/SR-13343). But for that issue a workaround exists. (Workaround is described in the thread)

When using a xcframework with spm, the xcframework needs to be included in every release. This would require a build for each release. Building the xcframework could be added in the archive script https://github.com/AzureAD/microsoft-authentication-library-for-objc/blob/dev/ReleaseArchive.sh. As a result the script would take more time to finish.

**Comment by Daekou** [comment]
Very interested in SPM support. Any updates ?

**Comment by honkmaster** [comment]
Any update?

**Comment by kaisong1990** [comment]
We are still looking into it, but no ETA at this point

**Comment by bengovernment** [comment]
My team has also migrated multiple dependencies from Carthage to SPM. MSAL remains the last framework we can't move.

**Comment by jontolof** [comment]
Hi. We are also eager to migrate this framework to SPM. Any news?

**Comment by oldalton** [comment]
@ameyapat is actively investigating this right now. While we still don't have an ETA, we'll keep you posted on the progress.

**Comment by crspybits** [comment]
@draskovits Do you have your XCFramework downloadable? Or a procedure to create it from this repo? Thanks!

**Comment by crspybits** [comment]
My initial step in making an XCFramework is to build MSAL as a framework. But this fails. E.g., When I try:

```
FRAMEWORK_NAME="MSAL (iOS Framework)"
SIMULATOR_ARCHIVE_PATH=MSignIn.Simulator.xcarchive

xcodebuild archive -scheme "${FRAMEWORK_NAME}" -destination="iOS Simulator" -archivePath "${SIMULATOR_ARCHIVE_PATH}" -sdk iphonesimulator SKIP_INSTALL=NO SUPPORTS_MACCATALYST=NO BUILD_LIBRARIES_FOR_DISTRIBUTION=YES
```

from the root directory of the repository, I get:

```
xcodebuild archive -scheme "${FRAMEWORK_NAME}" -destination="iOS Simulator" -archivePath "${SIMULATOR_ARCHIVE_PATH}" -sdk iphonesimulator SKIP_INSTALL=NO SUPPORTS_MACCATALYST=NO BUILD_LIBRARIES_FOR_DISTRIBUTION=YES
Command line invocation:
    /Applications/Xcode-12.3.app/Contents/Developer/usr/bin/xcodebuild archive -scheme "MSAL (iOS Framework)" "-destination=iOS Simulator" -archivePath MSignIn.Simulator.xcarchive -sdk iphonesimulator SKIP_INSTALL=NO SUPPORTS_MACCATALYST=NO BUILD_LIBRARIES_FOR_DISTRIBUTION=YES

User defaults from command line:
    destination = iOS Simulator
    IDEArchivePathOverride = /Users/chris/Desktop/microsoft-authentication-library-for-objc/MSignIn.Simulator.xcarchive

Build settings from command line:
    BUILD_LIBRARIES_FOR_DISTRIBUTION = YES
    SDKROOT = iphonesimulator14.3
    SKIP_INSTALL = NO
    SUPPORTS_MACCATALYST = NO

note: Using new build system
note: Building targets in parallel
note: Planning build
note: Constructing build description
error: /Users/chris/Desktop/microsoft-authentication-library-for-objc/MSAL/xcconfig/msal__common.xcconfig:1: could not find included file '../IdentityCore/IdentityCore/xcconfig/identitycore__common.xcconfig' in search paths (in target 'MSAL (iOS Framework)' from project 'MSAL')
error: /Users/chris/Desktop/microsoft-authentication-library-for-objc/MSAL/xcconfig/msal__release.xcconfig:2: could not find included file '../IdentityCore/IdentityCore/xcconfig/identitycore__release.xcconfig' in search paths (in target 'MSAL (iOS Framework)' from project 'MSAL')
error: /Users/chris/Desktop/microsoft-authentication-library-for-objc/MSAL/xcconfig/msal__common__ios.xcconfig:1: could not find included file '../IdentityCore/IdentityCore/xcconfig/identitycore__common__ios.xcconfig' in search paths (in target 'MSAL (iOS Framework)' from project 'MSAL')
error: /Users/chris/Desktop/microsoft-authentication-library-for-objc/MSAL/xcconfig/msal__common.xcconfig:1: could not find included file '../IdentityCore/IdentityCore/xcconfig/identitycore__common.xcconfig' in search paths (in target 'MSAL (iOS Framework)' from project 'MSAL')
error: /Users/chris/Desktop/microsoft-authentication-library-for-objc/MSAL/xcconfig/msal__release.xcconfig:2: could not find included file '../IdentityCore/IdentityCore/xcconfig/identitycore__release.xcconfig' in search paths (in target 'MSAL (iOS Framework)' from project 'MSAL')
error: /Users/chris/Desktop/microsoft-authentication-library-for-objc/MSAL/xcconfig/msal__common__ios.xcconfig:1: could not find included file '../IdentityCore/IdentityCore/xcconfig/identitycore__common__ios.xcconfig' in search paths (in target 'MSAL (iOS Framework)' from project 'MSAL')
error: /Users/chris/Desktop/microsoft-authentication-library-for-objc/MSAL/xcconfig/msal__common.xcconfig:1: could not find included file '../IdentityCore/IdentityCore/xcconfig/identitycore__common.xcconfig' in search paths (in target 'MSAL (iOS Framework)' from project 'MSAL')
error: /Users/chris/Desktop/microsoft-authentication-library-for-objc/MSAL/xcconfig/msal__release.xcconfig:2: could not find included file '../IdentityCore/IdentityCore/xcconfig/identitycore__release.xcconfig' in search paths (in target 'MSAL (iOS Framework)' from project 'MSAL')
error: /Users/chris/Desktop/microsoft-authentication-library-for-objc/MSAL/xcconfig/msal__common__ios.xcconfig:1: could not find included file '../IdentityCore/IdentityCore/xcconfig/identitycore__common__ios.xcconfig' in search paths (in target 'MSAL (iOS Framework)' from project 'MSAL')
warning: Traditional headermap style is no longer supported; please migrate to using separate headermaps and set 'ALWAYS_SEARCH_USER_PATHS' to NO. (in target 'MSAL (iOS Framework)' from project 'MSAL')
```

**Comment by wenjie-hou** [comment]
@draskovits Hey, I just made my own fork, and the xcframework build (https://github.com/wenjie-hou/microsoft-authentication-library-for-objc). It works for me with SPM

**Comment by crspybits** [comment]
Thanks for this, @wenjie-hou -- could we please have the commands you used to generate this?

**Comment by wenjie-hou** [comment]
> Thanks for this, @wenjie-hou -- could we please have the commands you used to generate this?

I use the same command as yours. Maybe you didn't clone the repo with `--recursive`?

**Comment by draskovits** [comment]
yeah basically did the same, but once for iPhone simulator and once for iPhoneSDK. 

Currently I use my fork, but the version is out date (I think it is 1.1.9) 
https://github.com/draskovits/microsoft-authentication-library-for-objc/tree/spm-support

**Comment by gao-sun** [comment]
For those who are struggling on waiting the next release, you can try [the forked version](https://github.com/scdt-china/microsoft-authentication-library-for-objc) **(iOS only)** which is based on the latest `dev` branch (the last version is 1.1.13). I added `rc1` suffix so do not forget to choose `branch` for resolving. Cheers.

![image](https://user-images.githubusercontent.com/14722250/105204941-a4df3d80-5b7f-11eb-883d-ed6c833165a8.png)

**Comment by danielelkington** [comment]
1.1.14 has been released now, but it doesn't seem to fix the problem - now there's a checksum issue.
![image](https://user-images.githubusercontent.com/20923693/105559740-0eb93c00-5d66-11eb-834e-f1b5897fbdfd.png)

**Comment by ameyapat** [comment]
@danielelkington mistakenly had tagged the wrong commit in the release. Can you retry?

**Comment by danielelkington** [comment]
This is working now - I think this issue can be closed.

**Comment by ameyapat** [comment]
Using MSAL as a swift package dependency is available in release 1.1.14. To use it as a package dependency please follow : https://github.com/AzureAD/microsoft-authentication-library-for-objc#using-swift-packages

Please let us know if there are any issues regarding this and thanks for your patience!

**Comment by zntfdr** [comment]
@ameyapat (and the rest of the MSAL team) thank you for adding SPM support! 👏🏻 

Everything works great while debugging, however archiving now fails with the following error:

```
error: exportArchive: IPA processing failed

Error Domain=IDEFoundationErrorDomain Code=1 "IPA processing failed" UserInfo={NSLocalizedDescription=IPA processing failed}
```

The only change I did is removing the MSAL pod and add the dependency via SPM following the linked guide: 
https://github.com/AzureAD/microsoft-authentication-library-for-objc#using-swift-packages

Do you have any leads on this? Thank you in advance 🙏🏻

**Comment by ChrisPearce** [comment]
Thanks for adding SPM support.

I can compile successfully but I'm encountering the following error when submitting the package to the App Store:
`2021-01-26T03:06:24.0949520Z [03:06:24]: WARNING ITMS-90806: "CFBundleIdentifier collision. Each bundle must have a unique bundle identifier. The bundle identifier 'com.microsoft.MSAL' is used in the bundles '[Payload/MyApp.app/Frameworks/MSAL.framework, Payload/MyApp.app/PlugIns/MSAL.framework]'"
`

**Comment by ameyapat** [comment]
@ChrisPearce Looks like a SPM bug where frameworks get copied under Frameworks and Plugins when building as per : https://forums.swift.org/t/swift-package-binary-framework-issue/41922/3. I guess for now the workaround is to have a script to remove the framework when building : https://forums.swift.org/t/swift-package-binary-framework-issue/41922/3

**Comment by ameyapat** [comment]
@zntfdr What do the distribution logs say when IPA processing fails? Maybe the .app has MSAL.framework embedded at multiple locations causing this issue. Could you try the same workaround as I mentioned in above comment : https://forums.swift.org/t/swift-package-binary-framework-issue/41922/3

**Comment by zntfdr** [comment]
@ameyapat Thank you! Adding the script phase works! 🎉 

For other people with this issue, here's the complete script phase that needs to be added:
<img width="907" alt="Screenshot 2021-01-27 at 09 52 28" src="https://user-images.githubusercontent.com/5277837/105936232-e000d580-6085-11eb-8b9a-53f77086aaf2.png">

Explanation:
1. add the following script to cleanup extra frameworks from your app plugins
```shell
#!/bin/bash

COUNTER=0
while [ $COUNTER -lt "${SCRIPT_INPUT_FILE_COUNT}" ]; do
    tmp="SCRIPT_INPUT_FILE_$COUNTER"
    FILE=${!tmp}

    echo "Removing $FILE"
    rm -rf "$FILE"
    let COUNTER=COUNTER+1
done
```

2. check `For install builds only` to run this script only when archiving (I know, the name is 🤷🏻‍♂️), this script is not needed when debugging
3. add `$(BUILT_PRODUCTS_DIR)/$(PLUGINS_FOLDER_PATH)/MSAL.framework` in the script phase Inputs, this is the framework that will be removed when the script is run.


This is a known SPM bug, it's even mentioned in the [Xcode 12.4 release notes](https://developer.apple.com/documentation/xcode-release-notes/xcode-12_4-release-notes/):
<img width="790" alt="Screenshot 2021-01-27 at 09 54 40" src="https://user-images.githubusercontent.com/5277837/105936218-d9725e00-6085-11eb-9797-2db7508a589e.png">

This is a temporary workaround that should be removed once fixed in Xcode (fingers crossed for Xcode 12.5 🤞🏻)

**Comment by cassianodialpad** [comment]
I'm facing a different issue here. It works fine running on simulator, but then I get the error below when trying to run on device. I have a workspace with a framework as an umbrella for SPM dependencies, and it works fine with all other dependencies I have. I was using MSAL with carthage before, and no issues. This happens only when adding MSAL to my SPM dependencies.

Worth noting that it looks like all other dependencies are static libraries embedded to my framework, but MSAL is being copied as a dynamic framework to my framework's `Frameworks` folder. I tried to delete it as per the workaround mentioned on Xcode 12.4 release notes, but no luck.

```
dyld: Library not loaded: @rpath/MSAL.framework/MSAL
  Referenced from: /private/var/containers/Bundle/Application/58E5E538-CD2D-45E2-970B-A8883EB56CFC/Dialpad.app/Frameworks/SDPCore.framework/SDPCore
  Reason: no suitable image found.  Did find:
	/private/var/containers/Bundle/Application/58E5E538-CD2D-45E2-970B-A8883EB56CFC/Dialpad.app/Frameworks/SDPCore.framework/Frameworks/MSAL.framework/MSAL: code signature in (/private/var/containers/Bundle/Application/58E5E538-CD2D-45E2-970B-A8883EB56CFC/Dialpad.app/Frameworks/SDPCore.framework/Frameworks/MSAL.framework/MSAL) not valid for use in process using Library Validation: mapped file has no cdhash, completely unsigned? Code has to be at least ad-hoc signed.
```

**Comment by guidedways** [comment]
I'm facing a different issue altogether - it does not seem to work for a macOS app. I see a `CodeSign Failed` build time error:

```
CodeSign /Users/MyUser/Library/Developer/Xcode/DerivedData/MyApp-dfxaoncuargnurehqwjntcnqwbzd/Build/Products/Debug/MyApp.app (in target 'MyApp' from project 'MyApp')
    cd /Users/MyUser/Development/git/BusyApps/calendar
    export CODESIGN_ALLOCATE\=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/codesign_allocate
    
Signing Identity:     "Apple Development: XXXXXXX"
Provisioning Profile: "Mac Team Provisioning Profile: com.busymac.busycal3"
                      (XXXXXXX)

    /usr/bin/codesign --force --sign 90EA95F14F419C329E568F32D3F68026EFC61BB1 --entitlements /Users/MyUser/Library/Developer/Xcode/DerivedData/MyApp-dfxaoncuargnurehqwjntcnqwbzd/Build/Intermediates.noindex/BusyCal.build/Debug/BusyCal\ Trial.build/MyApp.app.xcent --timestamp\=none /Users/MyUser/Library/Developer/Xcode/DerivedData/MyApp-dfxaoncuargnurehqwjntcnqwbzd/Build/Products/Debug/MyApp.app

/Users/MyUser/Library/Developer/Xcode/DerivedData/MyApp-dfxaoncuargnurehqwjntcnqwbzd/Build/Products/Debug/MyApp.app: code object is not signed at all
In subcomponent: /Users/MyUser/Library/Developer/Xcode/DerivedData/MyApp-dfxaoncuargnurehqwjntcnqwbzd/Build/Products/Debug/BusyCal.app/Contents/PlugIns/MSAL.framework
Command CodeSign failed with a nonzero exit code
```

**Comment by ameyapat** [comment]
@guidedways Your issue might be related to [this SPM issue](https://forums.swift.org/t/swiftpm-binarytarget-dependency-and-code-signing/38953). I would suggest the following : In your project's target -> Build Phases

- Add Copy Files phase with destination as Frameworks
- After that add Run script phase with `rm -rf "${CODESIGNING_FOLDER_PATH}/Contents/Frameworks/MSAL.framework"` (Similar to [this suggestion](https://forums.swift.org/t/swiftpm-binarytarget-dependency-and-code-signing/38953/43) )

**Comment by ameyapat** [comment]
@cassianodialpad Not totally sure but your issue might be related to this SPM bug : https://bugs.swift.org/browse/SR-13366

**Comment by guidedways** [comment]
@ameyapat That worked when I tried building the project, but fails when I try running the app under the debugger with the same error. I've give up for now as this seems far more painful than simply adding a direct dependency to Xcode.

**Comment by cassianodialpad** [comment]
> @cassianodialpad Not totally sure but your issue might be related to this SPM bug : https://bugs.swift.org/browse/SR-13366

@ameyapat Yes, probably... I was able to work around it by codesigning my internal umbrella framework and then moving MSAL to my app's `Frameworks` folder with this build post action:

`mv "${TARGET_BUILD_DIR}/${PRODUCT_NAME}.app/Frameworks/Umbrella.framework/Frameworks/MSAL.framework" "${TARGET_BUILD_DIR}/${PRODUCT_NAME}.app/Frameworks/MSAL.framework"`

**Comment by ameyapat** [comment]
Closing this issue. Will highlight some of these workaround in README. If there are any issues related to MSAL swift package dependency, please open a new issue. Thanks!

**Comment by guidedways** [comment]
@ameyapat We're still seeing issues trying to make this work for a mac app - the workaround does not help.

**Comment by ameyapat** [comment]
@guidedways The issue seems to caused because of swift package manager bug and not related to MSAL. The workaround provided had worked with our [sample app](https://github.com/Azure-Samples/ms-identity-macOS-swift-objc) when I tried it. Can you open a separate issue providing more details about the issue that you are facing?

**Comment by mthielemann** [comment]
> I also strangely see a `Current` folder (which should really be a symbolic link) in the `MSAL.framework` folder:
> 
> <img alt="Screenshot 2021-01-29 at 20 04 03@2x" width="363" src="https://user-images.githubusercontent.com/202437/106298294-29f7e000-626d-11eb-9ad6-fd70a8c9118e.png">

I see the exact same issue. Could make it work by doing this in a Run Script build phase:
```
rm -rf "${CODESIGNING_FOLDER_PATH}/Contents/Frameworks/MSAL.framework/"{Headers,Modules,MSAL,Resources}
rm -rf "${CODESIGNING_FOLDER_PATH}/Contents/Frameworks/MSAL.framework/Versions/Current"
rm -rf "${CODESIGNING_FOLDER_PATH}/Contents/Frameworks/MSAL.framework/Versions/A/"{Headers,Modules}
cd "${CODESIGNING_FOLDER_PATH}/Contents/Frameworks/MSAL.framework/Versions"
ln -s "A" "Current"
cd "${CODESIGNING_FOLDER_PATH}/Contents/Frameworks/MSAL.framework/"
ln -s "Versions/Current/Resources" "."
ln -s "Versions/Current/MSAL" "."
```

**Comment by guidedways** [comment]
> @ameyapat https://pspdfkit.com/guides/ios/10.2/knowledge-base/library-not-found-swiftpm/ for v10.2 worked. Thanks !

Did not work for me.

**Comment by WouterStraver** [comment]
I'm currently facing the exact same issues as @guidedways. It works fine on iOS but it does not work with MacOS. Even with the workarounds mentioned above.

**Comment by phil1995** [comment]
Is it possible that you also publish an App-Extension-Safe target like you did for CocoaPods?
Because currently when including it in an app extension the following warning appears:
`ld: warning: linking against a dylib which is not safe for use in application extensions: /Library/Developer/Xcode/DerivedData/XXX/Build/Products/Debug-iphonesimulator/MSAL.framework/MSAL`

**Comment by schlingding** [comment]
Did Xcode 13 break the workaround here? After having this fix for almost a year we're back to this error.

```
Command CodeSign failed with a nonzero exit code
```

We also get the warning
```
Couldn't resolve framework symlink for '/Users/user/Library/Developer/Xcode/DerivedData/ExampleApp/SourcePackages/artifacts/MSAL/MSAL.xcframework/macos-arm64_x86_64/MSAL.framework/Versions/Current': readlink(/Users/user/Library/Developer/Xcode/DerivedData/ExampleApp/SourcePackages/artifacts/MSAL/MSAL.xcframework/macos-arm64_x86_64/MSAL.framework/Versions/Current): Invalid argument (22)
```

**Comment by guidedways** [comment]
Just a remark - recently Google added SPM support to Firebase. If they can get firebase to play ball with SPM with zero tinkering, why can't MSAL?

**Comment by schlingding** [comment]
This may be due to the use of `binaryTarget`

**Comment by ferantivero** [comment]
@guidedways after following the workaround in [here](https://github.com/AzureAD/microsoft-authentication-library-for-objc/issues/737#issuecomment-770056675) with SPM + Xcode 13.2.1 + Swift 5.5 I've just been able to get a `Build Succeeded` but it is sort of unstable :)

First time you attempt to build, you will get the `CodeSign` error that ends up with a `Build  Failed` as you/others and I were experiencing, but then if you go for a second attempt "without cleaning the build folder` you get it successfully built. 

:eyes: @ameyapat the order of the added build phases won't affect this behavior.

**Comment by FooterManDev** [comment]
I'm getting this issue. No matter what workaround I do it isn't fixing the problem. The workaround phases all seem to execute after the MSAL error.
